src/clusterfuzz/_internal/bot/tasks/utasks/regression_task.py

```diff
@@ -182,52 +181,76 @@ def found_regression_near_extreme_revisions(
       return error
 
     if not is_crash:
-      regression_task_output.regression_range_start = revision_list[
-          current_index]
-      regression_task_output.regression_range_end = last_known_crashing_revision
-      return uworker_msg_pb2.Output(
-          regression_task_output=regression_task_output)
+      # We've found the latest passing revision, no need to binary search.
+      output.regression_range_start = revision
+      output.regression_range_end = last_known_crashing_revision
+      return uworker_msg_pb2.Output(regression_task_output=output)
 
-    last_known_crashing_revision = revision_list[current_index]
+    last_known_crashing_revision = revision
 
+  # All most recent revisions crash.
+  return None
+
+
+def check_earliest_revisions(
+    testcase: data_types.Testcase,
+    testcase_file_path: str,
+    job_type: str,
+    revision_range: List[int],
+    fuzz_target: Optional[data_types.FuzzTarget],
+    regression_task_output: uworker_msg_pb2.RegressionTaskOutput,
+) -> Optional[uworker_msg_pb2.Output]:
+  """Check that the earliest good build does not crash.
+
+  Adds information about any bad builds encountered while running to
+  `regression_task_output.build_data_list`.
+
+  Returns:
+    None if one of the earliest builds is good and does not crash.
+    An output proto if:
+
+    a. The earliest good build crashes, in which case the regression range is
+       set to [0, min_good_revision).
+    b. An error occurred.
+  """
   # Test to see if we crash in the oldest revision we can run. This is a pre-
   # condition for our binary search. If we do crash in that revision, it
   # implies that we regressed between the first commit and our first revision,
   # which we represent as 0:|min_revision|.
-  for _ in range(EXTREME_REVISIONS_TO_TEST):
-    min_revision = revision_list[min_index]
 
-    crashes_in_min_revision, error = _testcase_reproduces_in_revision(
+  revision_range = revision_range[:EXTREME_REVISIONS_TO_TEST]
+  for revision in revision_range:
+    is_crash, error = _testcase_reproduces_in_revision(
         testcase,
         testcase_file_path,
         job_type,
-        min_revision,
+        revision,
         regression_task_output,
         fuzz_target,
         should_log=False)
+
     if error:
+      # Skip bad build errors only.
       if error.error_type == uworker_msg_pb2.REGRESSION_BAD_BUILD_ERROR:
-        # If we find a bad build, potentially try another.
-        if min_index + 1 >= max_index:
```

Comment:
I think a subtle issue we're no longer handling is when the revision_range is small enough to have an overlap between `revision_range[:EXTREME_REVISIONS_TO_TEST]` and `revision_range[-EXTREME_REVISIONS_TO_TEST - 1:-1]`
See also line 169 in the initial file

Reply:
Thanks for pointing it out! I think the handling has not changed, actually.

On line 171 in the new file:

```py
  for revision in reversed(revision_range[-EXTREME_REVISIONS_TO_TEST - 1:-1])
```

This will return a slice with less than `EXTREME_REVISIONS_TO_TEST` elements if `revision_range` had less than `EXTREME_REVISIONS_TO_TEST + 1` elements in the first place. I think that's the same handling as line 169 in the original file.

Similarly for `checker_earliest_revisions()`, we do:

```py
  revision_range = revision_range[:EXTREME_REVISIONS_TO_TEST]
  for revision in revision_range:
```

Which will iterate over less than `EXTREME_REVISIONS_TO_TEST` revisions if `revision_range` is too small, just like it was handled on line 211 originally.

Reply:
```sh
$ python3            
[...]                                                                                                                     
>>> [1, 2, 3][-4:-1]                                                                                                            
[1, 2]
```